### PR TITLE
Migrate IntersectionObserver v2 article

### PR DESCRIPTION
**Pull request by tomayac**
Changes proposed in this pull request:

- Migrate https://developers.google.com/web/updates/2019/02/intersectionobserver-v2 to `/intersectionobserver-v2/`.

### CODE REVIEW

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
+---
+title: Trust is Good, Observation is Better—Intersection Observer v2
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:2]:
```suggestion
title: "Trust is good, observation is better: Intersection Observer v2"
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+subhead: |
+  Intersection Observer v2 adds the capability to not only observe intersections per se, but to also
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:4]: Is "v2" how this is styled on the spec itself? I am unable to find the actual spec.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+hero: image/8WbTDNrhLsU0El80frMBGE4eMCD3/wUCO2CgK5UOz6URpXEtd.jpeg
+alt: Eye of a cat looking through a gap in a fence.
+---
+
+Intersection Observer&nbsp;v1 is one of those APIs that's probably universally loved, and, now that
+[Safari supports it](https://webkit.org/blog/8582/intersectionobserver-in-webkit/) as well,
+it's also finally universally usable in all major browsers. For a quick refresher of the API,
+I recommend watching [Surma](../../../resources/contributors/surma)'s
+[Supercharged Microtip](https://www.youtube.com/embed/kW_atFXMG98) on Intersection
+Observer&nbsp;v1—also embedded below for your viewing pleasure—or reading Surma's in-depth
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:25]: Since this is an old article, I'm skipping a lot of writing feedback I would normally give. This part of the sentence loses me completely.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+[article](../../2016/04/intersectionobserver).
+People have used Intersection Observer&nbsp;v1 for a wide range of use cases like
+[lazy loading of images and videos](../../../fundamentals/performance/lazy-loading-guidance/images-and-video/),
+[being notified when elements reach `position: sticky`](../../2017/09/sticky-headers),
+[fire analytics events](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/visibility-manager.js),
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:30]:
```suggestion
[firing analytics events](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/visibility-manager.js),
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+const observer = new IntersectionObserver(onIntersection);
+observer.observe(document.querySelector('#some-target'));
+```
+
+## What's challenging with Intersection Observer&nbsp;v1?
+
+To be clear, Intersection Observer&nbsp;v1 is great, but it's not perfect. There are
+some corner cases where the API falls short. Let's have a closer look!
+The Intersection Observer&nbsp;v1 API can perfectly tell you when an element is scrolled into the
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:57]:
```suggestion
The Intersection Observer&nbsp;v1 API can tell you when an element is scrolled into the
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+window's viewport, but it *doesn't* tell you whether the element is covered
+by any other page content (that is, when the element is occluded) or whether
+the element's visual display has been modified by visual effects like `transform`, `opacity`,
+`filter`, etc., which *effectively* can make it invisible.
+
+Now while for an element in the top-level document this information can be determined by analyzing
+the DOM via JavaScript, for example via
+[`DocumentOrShadowRoot.elementFromPoint()`](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/elementFromPoint)
+and then digging deeper, the same information cannot be obtained if the element in question is
+located in a third-party iframe.
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:67]: I don't understand this sentence at all.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+## Why is actual visibility such a big deal?
+
+The Internet is, unfortunately, a place that attracts bad actors with even worse intentions.
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:71]:
```suggestion
The Internet is, unfortunately, a place that attracts bad actors with worse intentions.
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+to trick people into clicking their ads to increase the publisher's ad payout (at least
+for a short period, until the ad network catches them).
+Typically, such ads are served in iframes.
+Now if the publisher wanted to get users to click such ads, they could make the ad iframes
+completely transparent by applying a CSS rule `iframe { opacity: 0; }` and overlaying the iframes
+on top of something attractive, like a cute cat video that users would actually want to click.
+This is called *clickjacking*.
+You can see such a clickjacking attack in action in the upper section of this
+[demo](https://trick-ad-click.glitch.me/) (try "watching" the 🐈 cat video
+and ☑️ activate "trick mode").
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:82]:
```suggestion
[demo](https://trick-ad-click.glitch.me/) (try "watching" the cat video
and activate "trick mode").
```

Comment (jpmedley) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:82]: These emojis don't seem like they would pass the screen reader test. ("Try watching the cat cat video.")

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+You will notice that the ad in the iframe "thinks" it received legitimate clicks, even if it was
+completely transparent when you (pretendedly involuntarily) clicked it.
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:84]:
```suggestion
completely transparent when you (pretend-involuntarily) clicked it.
```

Comment (jpmedley) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:84]: I don't know of a grammar rule that actually applies here, but my suggestion would sound more correct to Anglophone ears.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/oSQhudX6lenD4Pld7PJD.png", alt="Tricking a user into clicking an ad by styling it transparent and overlaying it on top of something attractive.", width="800", height="533" %}
+
+## How does Intersection Observer&nbsp;v2 fix this?
+
+Intersection Observer&nbsp;v2 introduces the concept of tracking the actual "visibility" of a target
+element as a human being would define it.
+By setting an option in the
+[`IntersectionObserver` constructor](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver),
+intersecting
+[`IntersectionObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry)s
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:95]: I don't understand what's after the comma.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+and has no visual effects applied that would alter or distort its display on screen.
+In contrast, a `false` value means that the implementation cannot make that guarantee.
+
+An important detail of the
+[spec](https://w3c.github.io/IntersectionObserver/v2/#calculate-visibility-algo)
+is that the implementation *is permitted* to report *false negatives* (that is, setting `isVisible`
+to `false` even when the target element is completely visible and unmodified).
+For performance or other reasons, implementations should limit themselves to working with bounding
+boxes and rectilinear geometry; they shouldn't try to achieve pixel-perfect results for
+modifications like `border-radius`.
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:108]: >implementations should limit themselves

Browser implementations or uses of the API? Since this has been around for a while, we should know what browsers actually did rather than speculating on what they might do.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+That said, *false positives* are *not permitted* under any circumstances (that is, setting
+`isVisible` to `true` when the target element is not completely visible and unmodified).
+
+{% Aside 'warning' %}
+  Visibility is *much more expensive* to compute than intersection. For that reason,
+  Intersection Observer&nbsp;v2 is *not intended to be used broadly* in the way that
+  Intersection Observer&nbsp;v1 is. Intersection Observer&nbsp;v2 is focused on combatting fraud
+  and should be used only when Intersection Observer&nbsp;v1 functionality is *truly* insufficient.
+{% endAside %}
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:118]: This question feels like scope creep, but I'm going to ask it anyway.

What does "*truly* insufficient" mean? Seems like there's some additional guidance to be had in that question.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+## What does the new code look like in practice?
+
+The `IntersectionObserver` constructor now takes two additional configuration properties: `delay`
+and `trackVisibility`.
+The `delay` is a number indicating the minimum delay in milliseconds between notifications from
+the observer for a given target.
+The `trackVisibility` is a boolean indicating whether the observer will track changes in a target's
+visibility.
+
+⚠️ It's important to note here that when `trackVisibility` is `true`, `delay` is required to be at
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:129]: At the very least, remove this symbol. Possibly turn this paragraph into a note.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+least `100` (that is, no more than one notification every 100ms).
+As noted before, visibility is expensive to calculate, and this requirement is a precaution against
+performance degradation (and battery consumption). The responsible developer will use the
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:132]:
```suggestion
performance degradation and battery consumption. The responsible developer will use the
```

Comment (jpmedley) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:132]: That doesn't seem like a parenthetical issue. Why wasn't it mentioned, at least in passing, with the earlier reference to performnance?

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+*largest tolerable value* for delay.
+
+According to the current
+[spec](https://w3c.github.io/IntersectionObserver/v2/#calculate-visibility-algo), visibility is
+calculated as follows:
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:137]: In my search for the spec I draft I noticed that the v1 spec has a date of Feb 3, 2021. This one has a date of Nov 4, 2019. This might lead some people to think v2 is on its way to the grave. Do you know what's going on.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+- If the implementation cannot guarantee that the target is completely unoccluded by other page
+content, then the target is considered invisible.
+
+This means current implementations are pretty conservative with guaranteeing visibility.
+For example, applying an almost unnoticeable grayscale filter like `filter: grayscale(0.01%)`
+or setting an almost invisible transparency with `opacity: 0.99` would all render the element
+invisible.
+
+Below is a short code sample that illustrates the new API features. You can see this click tracking
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:159]:
```suggestion
Below is a short code sample that illustrates the new API features. You can see its click tracking
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+logic in action in the second section of the [demo](https://trick-ad-click.glitch.me/)
+(but now, try "watching" the 🐶 puppy video). Be sure to activate "trick mode" again to immediately
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:161]:
```suggestion
(but now, try "watching" the puppy video). Be sure to activate "trick mode" again to immediately
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+convert yourself into a shady publisher and see how Intersection Observer&nbsp;v2 prevents
+non-legitimate ad clicks from being tracked.
+This time, Intersection Observer&nbsp;v2 has our back! 🎉
+
+{% Aside %}
+  Different from typical lazy-loading code, if you use Intersection Observer to prevent this
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:167]:
```suggestion
  In contrast to typical lazy-loading code, if you use Intersection Observer to prevent this
```

Comment (jpmedley) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:167]: I know it's one more word than you used, but it's more idiomatic for English.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+  kind of clickjacking attacks, you *must not* `unobserve` the element after the first intersection.
```
Comment (jpmedley) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:168]:
```suggestion
  kind of clickjacking attack, you *must not* `unobserve` the element after the first intersection.
```

**Review by jpmedley — CHANGES_REQUESTED** [review]
I tried  not to go overboard, but I think there are things that must be addressed before republishing this.

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 13 above)
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:129]:
```suggestion
It's important to note here that when `trackVisibility` is `true`, `delay` is required to be at
```

Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:129]: Emoji be gone.

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 2 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:4]: It's indeed used like that: https://w3c.github.io/IntersectionObserver/v2/#ref-for-intersectionobserver⑤:~:text=V2.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+alt: Eye of a cat looking through a gap in a fence.
+---
+
+Intersection Observer&nbsp;v1 is one of those APIs that's probably universally loved, and, now that
+[Safari supports it](https://webkit.org/blog/8582/intersectionobserver-in-webkit/) as well,
+it's also finally universally usable in all major browsers. For a quick refresher of the API,
+I recommend watching [Surma](../../../resources/contributors/surma)'s
+[Supercharged Microtip](https://www.youtube.com/embed/kW_atFXMG98) on Intersection
+Observer&nbsp;v1—also embedded below for your viewing pleasure—or reading Surma's in-depth
+[article](../../2016/04/intersectionobserver).
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:26]:
```suggestion
I recommend watching [Surma](../../../resources/contributors/surma)'s
[Supercharged Microtip](https://www.youtube.com/embed/kW_atFXMG98) on Intersection
Observer&nbsp;v1 that is embedded below.
You can also read Surma's in-depth
[article](https://developers.google.com/web/updates/2016/04/intersectionobserver).
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 3 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:25]: Massively simplified. These horrible [intertwingled](https://www.urbandictionary.com/define.php?term=intertwingled) sentences…

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+window's viewport, but it *doesn't* tell you whether the element is covered
+by any other page content (that is, when the element is occluded) or whether
+the element's visual display has been modified by visual effects like `transform`, `opacity`,
+`filter`, etc., which *effectively* can make it invisible.
+
+Now while for an element in the top-level document this information can be determined by analyzing
+the DOM via JavaScript, for example via
+[`DocumentOrShadowRoot.elementFromPoint()`](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/elementFromPoint)
+and then digging deeper, the same information cannot be obtained if the element in question is
+located in a third-party iframe.
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:68]:
```suggestion
For an element in the top-level document, this information can be determined by analyzing
the DOM via JavaScript, for example via
[`DocumentOrShadowRoot.elementFromPoint()`](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/elementFromPoint)
and then digging deeper.
In contrast, the same information cannot be obtained if the element in question is
located in a third-party iframe.
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 6 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:68]: Split up into two sentences.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+
+## How does Intersection Observer&nbsp;v2 fix this?
+
+Intersection Observer&nbsp;v2 introduces the concept of tracking the actual "visibility" of a target
+element as a human being would define it.
+By setting an option in the
+[`IntersectionObserver` constructor](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver),
+intersecting
+[`IntersectionObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry)s
+(pardon the wrong plural ending here) will then contain a new boolean field named `isVisible`.
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:97]:
```suggestion
By setting an option in the
[`IntersectionObserver` constructor](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver),
intersecting
[`IntersectionObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry)
instances will then contain a new boolean field named `isVisible`.
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 10 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:96]: Yeah, it was weird. Rephrased.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+A `true` value for `isVisible` is a strong guarantee from the underlying implementation
+that the target element is completely unoccluded by other content
+and has no visual effects applied that would alter or distort its display on screen.
+In contrast, a `false` value means that the implementation cannot make that guarantee.
+
+An important detail of the
+[spec](https://w3c.github.io/IntersectionObserver/v2/#calculate-visibility-algo)
+is that the implementation *is permitted* to report *false negatives* (that is, setting `isVisible`
+to `false` even when the target element is completely visible and unmodified).
+For performance or other reasons, implementations should limit themselves to working with bounding
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:107]:
```suggestion
For performance or other reasons, browsers limit themselves to working with bounding
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+boxes and rectilinear geometry; they shouldn't try to achieve pixel-perfect results for
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:108]:
```suggestion
boxes and rectilinear geometry; they don't try to achieve pixel-perfect results for
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 11 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:109]: Yeah, this needed rephrasing. Browser implementations was the one.

File: src/site/content/en/blog/intersectionobserver-v2/index.md [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+modifications like `border-radius`.
+
+That said, *false positives* are *not permitted* under any circumstances (that is, setting
+`isVisible` to `true` when the target element is not completely visible and unmodified).
+
+{% Aside 'warning' %}
+  Visibility is *much more expensive* to compute than intersection. For that reason,
+  Intersection Observer&nbsp;v2 is *not intended to be used broadly* in the way that
+  Intersection Observer&nbsp;v1 is. Intersection Observer&nbsp;v2 is focused on combatting fraud
+  and should be used only when Intersection Observer&nbsp;v1 functionality is *truly* insufficient.
```
Comment (tomayac) [inline src/site/content/en/blog/intersectionobserver-v2/index.md:118]:
```suggestion
  and should be used only when the visibility information is needed and
  when Intersection Observer&nbsp;v1 functionality is therefore insufficient.
```

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 12 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:119]: I have made this clearer.

File: src/site/content/en/blog/intersectionobserver-v2/index.md (hunk 15 above)
Comment (tomayac) [reply src/site/content/en/blog/intersectionobserver-v2/index.md:138]: Yeah, spec work on V2 has halted for the moment in favor for work on V1, but V2 has not been abandoned.

**Comment by tomayac** [comment]
@jpmedley Thanks for the review, I have addressed your questions and remarks.

**Review by jpmedley — APPROVED** [review]
